crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -409,23 +424,38 @@ fn assert_component_access_compatibility(
     panic!("error[B0001]: Query<{}, {}> in system {system_name} accesses component(s) {accesses}in a way that conflicts with a previous system parameter. Consider using `Without<T>` to create disjoint Queries or merging conflicting Queries into a `ParamSet`. See: https://bevyengine.org/learn/errors/b0001", ShortName(query_type), ShortName(filter_type));
 }
 
+/// SAFETY: Caller must ensure no other mutable access to QueryState<D,F>
+unsafe fn get_query_state<'w, D: QueryData + 'static, F: QueryFilter + 'static>(
+    entity: Entity,
+    world: UnsafeWorldCell<'w>,
+) -> Option<&'w QueryState<D, F>> {
+    let state = world
+        .get_entity(entity)
+        .unwrap()
+        .get::<QueryState<D, F>>()
```

Comment:
We'll want to benchmark to make sure this extra lookup doesn't degrade performance.  

And we probably want to cache the `ComponentId` for `QueryState<D, F>` to avoid the hashtable lookup.